internal/app/machined/pkg/controllers/kubespan/manager.go

```diff
@@ -0,0 +1,391 @@
+	Close() error
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *ManagerController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      kubespan.ConfigType,
+			ID:        pointer.ToString(kubespan.ConfigID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: kubespan.NamespaceName,
+			Type:      kubespan.PeerSpecType,
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: kubespan.NamespaceName,
+			Type:      kubespan.IdentityType,
+			ID:        pointer.ToString(kubespan.LocalIdentity),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *ManagerController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.LinkSpecType,
+			Kind: controller.OutputShared,
+		},
+		{
+			Type: network.AddressSpecType,
+			Kind: controller.OutputShared,
+		},
+		{
+			Type: network.RouteSpecType,
+			Kind: controller.OutputShared,
+		},
+		{
+			Type: kubespan.PeerStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *ManagerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	var (
+		tickerC <-chan time.Time
+		ticker  *time.Ticker
+	)
+
+	if ctrl.WireguardClientFactory == nil {
+		ctrl.WireguardClientFactory = func() (WireguardClient, error) {
+			return wgctrl.New()
+		}
+	}
+
+	if ctrl.PeerReconcileInterval == 0 {
+		ctrl.PeerReconcileInterval = DefaultPeerReconcileInterval
+	}
+
+	wgClient, err := ctrl.WireguardClientFactory()
+	if err != nil {
+		return fmt.Errorf("error creating wireguard client: %w", err)
+	}
+
+	defer wgClient.Close() //nolint:errcheck
+
+	for {
+		var updateSpecs bool
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			updateSpecs = true
+		case <-tickerC:
+		}
+
+		cfg, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, kubespan.ConfigType, kubespan.ConfigID, resource.VersionUndefined))
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("error getting kubespan configuration: %w", err)
+		}
+
+		if cfg == nil || !cfg.(*kubespan.Config).TypedSpec().Enabled {
+			if ticker != nil {
+				ticker.Stop()
+
+				tickerC = nil
+			}
+
+			// KubeSpan is not enabled, cleanup everything
+			if err = ctrl.cleanup(ctx, r); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		if ticker == nil {
+			ticker = time.NewTicker(ctrl.PeerReconcileInterval)
+			tickerC = ticker.C
+		}
+
+		cfgSpec := cfg.(*kubespan.Config).TypedSpec()
+
+		localIdentity, err := r.Get(ctx, resource.NewMetadata(kubespan.NamespaceName, kubespan.IdentityType, kubespan.LocalIdentity, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting first MAC address: %w", err)
```

Comment:
`localIdentity !== first mac`?

Reply:
I mean, it _is_ "currently" the MAC, but the local code does not match that... it's simply the identity

Reply:
That's just wrong error message, copy-paste, I'll fix it, thanks